src/main/java/org/broadinstitute/barclay/argparser/CommandLineArgumentParser.java

```diff
@@ -179,32 +109,231 @@ public CommandLineArgumentParser(
+     * calling this method.
+     *
+     * @return The commandline, or null if {@link #parseArguments(PrintStream, String[])}
+     * hasn't yet been called, or didn't complete successfully.
+     */
+    @Override
+    public String getCommandLine() {
+        final StringBuilder commandLineString = new StringBuilder();
+        String tempString;
+        commandLineString.append(callerArguments.getClass().getSimpleName());
+
+        // positional args first
+        if( positionalArgumentDefinition != null) {
+            tempString = positionalArgumentDefinition.getCommandLineDisplayString();
+            commandLineString.append(tempString.length() > 0 ? " " + tempString : "");
+        }
+
+        // then args that were explicitly set
+        tempString = argumentDefinitions.stream()
+                .filter(argumentDefinition -> argumentDefinition.getHasBeenSet())
+                .map(NamedArgumentDefinition::getCommandLineDisplayString)
+                .collect(Collectors.joining(" "));
+        commandLineString.append(tempString.length() > 0 ? " " + tempString : "");
+
+        // finally, args that weren't explicitly set, but have a default value
+        tempString = argumentDefinitions.stream()
+                        .filter(argumentDefinition -> !argumentDefinition.getHasBeenSet() &&
+                                !argumentDefinition.getDefaultValueAsString().equals(NamedArgumentDefinition.NULL_ARGUMENT_STRING))
+                        .map(NamedArgumentDefinition::getCommandLineDisplayString)
+                        .collect(Collectors.joining(" "));
+        commandLineString.append(tempString.length() > 0 ? " " + tempString : "");
+
+        return commandLineString.toString();
+    }
+
+    /**
+     * A typical command line program will call this to get the beginning of the usage message,
+     * and then append a description of the program, like this:
+     * commandLineParser.getStandardUsagePreamble(getClass()) + "Frobnicates the freebozzle."
+     */
+    @Override
+    public String getStandardUsagePreamble(final Class<?> mainClass) {
+        final String preamble = "USAGE: " + mainClass.getSimpleName() + " [arguments]\n\n";
+        if (mainClass.getAnnotation(ExperimentalFeature.class) != null) {
+            return EXPERIMENTAL_PREFIX + preamble;
+        } else if (mainClass.getAnnotation(BetaFeature.class) != null) {
+            return BETA_PREFIX + preamble;
+        } else {
+            return preamble;
         }
+    }
 
-        this.programProperties = this.callerArguments.getClass().getAnnotation(CommandLineProgramProperties.class);
+    // Validate that any mutex targets exist. NOTE: it isn't a requirement for correct behavior that mutex
+    // arguments have symmetric declarations, and its not enforced. But they should be declared that way
+    // so that the relationship is reflected in usage/help/doc for all arguments involved.
+    private void validateArgumentDefinitions() {
```

Comment:
Resisting since although its currently only mutex validation, it may expand to be broader as we add new features. I update the comment saying that it currently only does mutex validation.

Reply:
Sounds good.